### PR TITLE
Pipe operator

### LINKED ISSUE #40: Provide a means to change the context node
Input

```
{
  "a": {
    "b": 1,
    "c": 2, 
    "d": 3
  } 
}
```

I want to simplify `[  .a.b, .a.c, a.d ]` to ` .a | [.b, .c, .d]`. The idea is that the context node `.` inside `[.b, .c, .d] `  no longer points to the input node but to `.a` instead.

### LINKED ISSUE #104: Object matching with subobjects
Hello,

Currently you can copy the keys and their values from the input object with "*: ."  What if the construction of the output object should be based on some subobject? Say you have the following input object:

`{
    "a": 0,
    "b": {
      "c": 1,
      "d": 2
    }
}`

The output should be this:

`{
    "c": 1,
    "d": 2
}`

It would be nice to be able to write the following to achieve this:

`{
    .b.*: .
}`

Best regards,

Werner.

### PR DESCRIPTION
**Pull request by ecerulm**
Add support for pipe operator as discussed in issue #40. 

This is Work In Progress, the basic works:

```
{"bar": "baz","foo":{ "a": "b" } | {"type" : "Anonymized-View",* : .}} -> {"bar":"baz","foo":{"type":"Anonymized-View","a":"b"}}` 
.a | [.b, .c, .d] -> [1,2,3]
```

What remains that I know of:

- [x] decide the operator precedence with corresponding test
- [x] decide the operator associativity with corresponding test
- [x] test interaction with for expressons, if expressions, object for expressions (on both sides of pipe operator)

Fixes #40
Fixes #104

### CODE REVIEW

**Comment by ecerulm** [comment]
Changed the pipe operator precedence to be higher that the logical operators (and/or) and the comparator operators (==, >, etc) but lower than the arithmetic opeators (+,-,*, etc) with corresponding tests

File: src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+  @Override
+  public JsonNode apply(Scope scope, JsonNode input) {
+    right.computeMatchContexts(new DotExpression(new Location(null, 0,0)));
+    return right.apply(scope, left.apply(scope,input));
+  }
+
+  @Override
+  public JsonNode perform(JsonNode v1, JsonNode v2) {
+    throw new RuntimeException("this should NOT be reachable");
```
Comment (larsga) [inline src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java]: You need to throw `JsltException` everywhere in JSLT. We should never throw anything else.

File: src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+public class PipeOperator extends AbstractOperator {
+
+  public PipeOperator(ExpressionNode left, ExpressionNode right,
+                      Location location) {
+    super(left, right, "|", location);
+  }
+
+  @Override
+  public JsonNode apply(Scope scope, JsonNode input) {
+    right.computeMatchContexts(new DotExpression(new Location(null, 0,0)));
```
Comment (larsga) [inline src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java]: We want to do this at compile time instead of repeating it every time the expression is evaluated. It's better to implement it in the `computeMatchContexts` method. Note that you need to do it for both `left` and `right`.

File: src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import com.schibsted.spt.data.jslt.JsltException;
+
+public class PipeOperator extends AbstractOperator {
+
+  public PipeOperator(ExpressionNode left, ExpressionNode right,
+                      Location location) {
+    super(left, right, "|", location);
+  }
```
Comment (larsga) [inline src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java]: This object is missing `optimize()`. I can't think of any optimization opportunities for the pipe operator, but we are now losing out on any optimization that could happen inside `left`  and  `right`

**Comment by larsga** [comment]
I'm very happy to see this PR, but also a bit sad. Happy because it's great to have more than one person who can contribute new features. Sad because I was looking forward to implementing this myself. :-) Anyway, let's go with this PR.

I'm on a work trip in Russia at the moment, so it will take a while to review this properly. I've put in a couple of comments, and will add more as I have time.

One thing that's clear is that this needs more tests. I'll try to contribute more test cases once I've had time to go through the tests more properly.

File: src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java (hunk 2 above)
Comment (larsga) [reply src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java]: Basically, the inherited logic in `computeMatchContexts` is correct for `left`, but you need this line to handle `right`

File: src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java (hunk 3 above)
Comment (ecerulm) [reply src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java]: I think since `PipeOperator` inherits from `AbstractOperator` the `AbstractOperator.optimize()` handles this. (by calling `left.optimize()` and `right.optimize` see [AbstractOperator.java](https://github.com/schibsted/jslt/blob/7274a94940500ba758a4eb1351d7036b933fc14f/src/main/java/com/schibsted/spt/data/jslt/impl/AbstractOperator.java#L50-L60))

File: src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java (hunk 1 above)
Comment (ecerulm) [reply src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java]: Done.

File: src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java (hunk 2 above)
Comment (ecerulm) [reply src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java]: Right, I addressed this in the last commit. Take a look at the latest one.

**Comment by ecerulm** [comment]
Yes, more test would be great. I have now pushed test for if-expressions, for-expressions, and object-for-expressions covering the basics. It would be great if you can come up with corner cases,etc. 

I also updated the tutorial to mention the pipe operator

File: src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java (hunk 3 above)
Comment (larsga) [reply src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java]: You're right. My bad. Sorry. Better to take the method out. (I keep forgetting about `AbstractOperator`.)

Comment (ecerulm) [reply src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java]: Done!

**Comment by ecerulm** [comment]
@larsga if you have any proposal on what other tests must be implemented you can share it here and I will implement them.

File: src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
+
+// Copyright 2018 Schibsted Marketplaces Products & Technology As
```
Comment (larsga) [inline src/main/java/com/schibsted/spt/data/jslt/impl/PipeOperator.java:2]: 2020, I guess? :)

File: src/test/resources/query-tests.yaml [hunk 5]
```diff
@@ -107,6 +107,226 @@ tests:
...
+          {
+            "type" : "Anonymized-View",
+             "a": "b"
+          }
+      }
+
+  - input:
+    query: |
+      1 == 3 | (. - 2)
+    output: true
```
Comment (larsga) [inline src/test/resources/query-tests.yaml:191]: I think this shows that the operator precedence is wrong. It seems to me that the pipe should have the highest precedence, so that this becomes `(1 == 3) | (. - 2)` and gives an error.

The same with the `or` example below. I think it should be `(true or false) | { "a" : .}` producing `{ "a" : true}`

What do you think?

Comment (ecerulm) [reply src/test/resources/query-tests.yaml:191]: > It seems to me that the pipe should have the highest precedence

You mean lowest precedence ,right? 

I can't remember what was the rationale behind keeping comparators and logical operators lower than the pipe operator, so for me it's ok to change it to lowest if it feels more natural.

Maybe we should have a doc page just listing the operator precedence in general as well.

Comment (larsga) [reply src/test/resources/query-tests.yaml:191]: > You mean lowest precedence ,right?

Argh. Yes, lowest. Never remember the terminology. :-)

> for me it's ok to change it to lowest if it feels more natural.

It felt that way to me. I say we try it, and once we have some experience with the master branch we can keep it or change it.

> Maybe we should have a doc page just listing the operator precedence in general as well.

Yeah, good point!

Comment (ecerulm) [reply src/test/resources/query-tests.yaml]: Ok, I made pipe operator the lowest precedence operator,

I created issue #128 to handle adding the documentation page with the operator precedence

**Comment by ecerulm** [comment]
I think 9e7ee83 addresses the operator precedence issue by settings pipe operator to have the lowest precedence. Double check the tests,
